### DISCUSSION TITLE
`getInstance` should work regardless of how I've created the object instance.

### DISCUSSION THREAD

**Discussion by crespire** [twbs/bootstrap#39311] [Q&A] [answered]
Currently, I have a few modals created via data attributes, but #32337 suggests that trying to use `bootstrap.Modal.getInstance()` will not work on such instances. I have tried as well in code with no such luck. Passing in the DOM element, or a string with the ID in both `#id` and `id` formats did not work, and returned `null` every time.

I have a bog standard modal but would like to modify some of its behaviour, especially around the `hide.bs.modal` event, and would like the option to pragrammatically close the modal under certain circumstances.

So, I ended up wanting to do this:

```js
const modal = bootstrap.Modal.getInstance('#data-attribute-modal');
const cancelButton = document.getElementById('cancel');

modal.addEventListener('hide.bs.modal', event => {
  if (!modal.querySelector('#form').checkValidity()) {
    event.preventDefault()
  }
});

cancelButton.addEventListner('click', event => {
  modal.hide();
});
```

In this case, I want to prevent the modal from closing if the form is invalid, but would like to close the modal regardless of the cancel button has been clicked.

Because my modal is built via data attributes, `bootstrap.Modal.getInstance('#data-attribute-modal')` will return `null` and I am unable to programmatically do this.

**Comment by julien-deramond** [comment]
Hey @crespire 
I've converted your feature request to a Q&A discussion for now.
I quickly tried something with [this CodePen](https://codepen.io/julien-deramond/pen/poqXGdr) and if I have well understood your issue, I probably managed to do what you wanted thanks to `getOrCreateInstance()`.

**Reply by crespire** [reply]
So, `getOrCreateInstance` works, except that my existing modal won't close when sent `modal.hide()`. I suspect that is because the active instance in the code (`const modal = bootstrap.Modal.getOrCreateInstance('#data-attribute-modal');`) is not actually the instance controlling the modal that opens.

Pardon all the edits, let me fork the codepen and play around with it.

**Reply by julien-deramond** [reply, marked as answer]
This [other variant of the same CodePen](https://codepen.io/julien-deramond/pen/oNJrmKv) creates a "Try me" button, when you hover out of this button, `modal.hide()` is called and closes the modal. Seems to work.
In the previous CodePen, `event.preventDefault();` was associated with `hide.bs.modal` event just to show that it was possible to intercept it. When `modal.hide()` was called, we entered into the `hide.bs.modal` callback, and `event.preventDefault()` stopped the propagation of the event, so it was impossible to close the modal anymore.

**Reply by crespire** [reply]
Thanks, @julien-deramond - in playing around with it, I see the problem is related to my validation logic which was always firing the `event.preventDefault()`. So using `getOrCreateInstance` works.

However, is there a reason why `getInstance` doesn't work in this case? It does seem to behave with `getOrCreateInstance`.

**Reply by julien-deramond** [reply]
Sorry @crespire, forgot to answer this question:

> However, is there a reason why `getInstance` doesn't work in this case? It does seem to behave with `getOrCreateInstance`.

The instance is only created after having clicked on the button which triggers the creation and the rendering of the modal. 
In the internal code of Bootstrap's modals:

```js
EventHandler.on(document, EVENT_CLICK_DATA_API, SELECTOR_DATA_TOGGLE, function (event) {
   // ...
   const data = Modal.getOrCreateInstance(target)
   // ...
}
```

It's not maybe the best event to attach it in my example, but just to show you the difference, here is [a CodePen getting the instance as you did but only when the modal is displayed (so after on the button's click event)](https://codepen.io/julien-deramond/pen/YzBXJEQ). This time, you can use `getInstance()` because it has already been created by Bootstrap when you clicked on the button.

```js
const triggerButton = document.getElementById('triggerButton').addEventListener('click', event => {
  const modal = bootstrap.Modal.getInstance('#data-attribute-modal');
  console.log(modal) // this is not `null` you can get the instance
}

const modal = bootstrap.Modal.getInstance('#data-attribute-modal');
console.log(modal) // this is `null` because the instance is not yet created when the page is loaded, nobody clicked on the button
```

**Reply by crespire** [reply]
Thanks, Julien! I appreciate the time you've taken to explain it to me. That makes perfect sense. I think this gives me an alternative way to handle the problem.
